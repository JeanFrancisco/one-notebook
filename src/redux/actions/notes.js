import Swal from 'sweetalert2';
import { addDoc, collection, deleteDoc, doc, getDocs, orderBy, updateDoc } from '@firebase/firestore';
import {
    NOTES_NEW_NOTE,
    NOTES_SET_ACTIVE,
    NOTES_DISPLAY_IN_SIDEBAR,
    NOTES_UPDATED_NOTE,
    NOTES_DROP_NOTE,
    NOTES_CLEANING_WHEN_LOGOUT,
    NOTES_CHANGE_ORDER,
} from '../constants/typesNotes';
import { uploadFile } from '../../helpers/files';
import { db as firestoreDB } from '../../config/firebase';

export const actionAddNewNote = () => {
    return async (dispatch, getState) => {
        const { uid: user_uid } = getState().auth;

        const newNote = {
            title: '',
            body: '',
            urlImage: '',
            createdAt: new Date().getTime(),
        };

        const doc = await addDoc( collection(firestoreDB, `${user_uid}/journal/notes`), newNote);

        const createdNote = {
            id: doc.id,
            ...newNote,
        };

        dispatch( actionNewNote( createdNote ) );
        dispatch( actionSetActiveNote( createdNote ) );
    }
};

export const actionNewNote = ( createdNote ) => ({
    type: NOTES_NEW_NOTE,
    payload: createdNote,
});

export const actionSetActiveNote = ( activeNote ) => ({
    type: NOTES_SET_ACTIVE,
    payload: activeNote,
});

export const actionSavingChanges = () => {
    return async (dispatch, getState) => {
        const user_uid = getState().auth.uid;
        const activeNote = getState().notes.active;

        const { id: note_id, ...noteForFirestore } = activeNote;
        await updateDoc( doc(firestoreDB, `${user_uid}/journal/notes/${ note_id }`), noteForFirestore )
            .then(() => {
                dispatch( refreshNote(activeNote) );

                Swal.fire({
                    title: 'Guardada exitosamente',
                    html: '<p><small>Su nota se guardó correctamente</small></p>',
                    icon: 'success',
                    timer: 1300,
                    showConfirmButton: true,
                    confirmButtonText: 'Aceptar',
                    confirmButtonColor: 'limegreen',
                });
            })
            .catch(reason => {
                console.log( reason.code );
            });
    }
}

export const refreshNote = ( updatedNote ) => ({
    type: NOTES_UPDATED_NOTE,
    payload: { ...updatedNote },
});

export const getFirebaseNotes = () => {
    return async ( dispatch, getState ) => {
        const { uid } = getState().auth;

        const snapNotes = await getDocs( collection(firestoreDB, `${uid}/journal/notes`) );
        const notes = [];

        snapNotes.forEach( snapHijo => {
            notes.push({
                id: snapHijo.id,
                ...snapHijo.data(),
            })
        });

        dispatch( displayNotes(notes) );
    }
}

export const displayNotes = ( notes ) => ({
    type: NOTES_DISPLAY_IN_SIDEBAR,
    payload: notes,
});

export const actionSetAscendingOrder = () => ({
    type: NOTES_CHANGE_ORDER,
    payload: {
        order: 'asc',
    }
});

export const actionSetDescendingOrder = () => ({
    type: NOTES_CHANGE_ORDER,
    payload: {
        order: 'desc',
    }
});

export const actionReorderByRule = ( orderBy ) => ({
    type: NOTES_CHANGE_ORDER,
    payload: {
        orderby: orderBy,
    }
});

export const actionStartUploadFile = (file) => {
    return async (dispatch, getState) => {
 
        Swal.fire({
            title: 'Subiendo archivos...',
            text: 'Por favor, espere un momento...',
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: () => {
                Swal.showLoading();
            },
        });

        const activeNote = getState().notes.active;
        const { secure_url } = await uploadFile(file);

        dispatch( actionSetActiveNote({ ...activeNote, urlImage: secure_url }) );
        await dispatch( actionSavingChanges() );

        Swal.close();
    }
}

export const actionStartDeletingNote = (note_id) => {
    return async (dispatch, getState) => {
        const user_uid = getState().auth.uid;

        try {
            const resp = await deleteDoc( doc( firestoreDB, `${user_uid}/journal/notes/${note_id}`) );

            dispatch( deleteNote( note_id ) );

            Swal.fire({
                icon: 'success',
                timer: 1250,
                title: 'Eliminado',
                text: 'Su nota fue eliminada',
                confirmButtonText: 'Aceptar',
                confirmButtonColor: 'green',
                showCloseButton: true,
            });

        } catch (e) {
            Swal.fire({
                icon: 'error',
                html: 'Error en la conexión con Firestore DB',
                title: 'Error. <small>No pudimos terminar la eliminación</small>',
                footer: `<small>For more details: ${e.code}</small>`,
                confirmButtonText: 'Entendido',
                confirmButtonColor: 'oragered', 
                showCloseButton: true,
            });

            console.error(e);
        }
    }
}

export const deleteNote = (noteID) => ({
    type: NOTES_DROP_NOTE,
    payload: noteID
});

export const cleanupNotesLogout = () => ({
    type: NOTES_CLEANING_WHEN_LOGOUT
});