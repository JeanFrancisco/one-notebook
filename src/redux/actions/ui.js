import { UI_FINISH_LOADING, UI_START_LOADING } from "../constants/typesUi";

export const actionStartLoading = () => ({
    type: UI_START_LOADING,
});

export const actionFinishLoading = () => ({
    type: UI_FINISH_LOADING,
});