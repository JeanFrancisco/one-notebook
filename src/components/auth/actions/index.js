import {
    createUserWithEmailAndPassword,
    signInWithEmailAndPassword,
    GoogleAuthProvider,
    signInWithPopup,
    updateProfile,
    getAuth,
    signOut,
} from "firebase/auth";
import Swal from 'sweetalert2';
import { app } from "../../../config/firebase";
import { cleanupNotesLogout } from "../../../redux/actions/notes";
import { actionFinishLoading, actionStartLoading } from "../../../redux/actions/ui";
import {
    AUTH_LOGIN,
    AUTH_LOGOUT,
    AUTH_SET_ERROR,
    AUTH_REMOVE_ERROR
} from "../constants";

export const authLogin = (uid, displayName) => ({
    type: AUTH_LOGIN,
    payload: {
        uid,
        displayName,
    }
});

export const authSignInWithGoogle = () => {
    return async (dispatch) => {

        dispatch( actionStartLoading() );

        const auth = getAuth( app );

        await signInWithPopup( auth, new GoogleAuthProvider() )
        .then( userCredentials => {
            // const { user: { displayName, uid } } = userCredentials;
            
            // dispatch( authLogin(uid, displayName) );
        })
        .catch( reason => {
            Swal.fire({
                icon: 'warning',
                html: '<small>No se pudo iniciar la sesión. <br/> <br/> ¿Alcanzó a ver si ocurrió algo: con el Pop-up de autenticación de Google?</small>',
                title: 'Oups! Lo sentimos.',
                iconHtml: '<i class="fa fa-warning"></i>',
                footer: `<small>Reason Code: ${reason.code}</small>`,
                showCloseButton: true,
                confirmButtonText: 'Aceptar',
            });
        })
        .finally(() => {
            dispatch( actionFinishLoading() );
        });
    };
}

export const authEmailPasswordSignIn = ({ email, password }) => {
    return async dispatch => {
        dispatch( actionStartLoading() );

        const auth = getAuth(app);

        await signInWithEmailAndPassword(auth, email, password)
            .then( ({ user }) => {
                // const { uid, displayName } = user;

                // dispatch( authLogin(uid, displayName) );
            }).catch( reason => {
                Swal.fire({
                    title: 'Oups! <small>Algo ocurrió mientras iniciabamos tu sesión.</small>',
                    html: `<p><b><small>¿Qué acciones puedes hacer?</small></b></p>
                        <ul style="text-align: left">
                            <li><small>Puedes revisar por favor tu email y contraseña.</small></li>
                            <li><small>También que te puedes conectar a Internet. En especial a los servicios de Google.</small></li>
                        </ul>`,
                    confirmButtonText: 'Entendido',
                    icon: 'warning',
                    iconColor: 'orange',
                    showCloseButton: true,
                    confirmButtonColor: 'orangered',
                    footer: `<small>Reason Code: ${reason.code}</small>`,
                })
            }).finally(() => {
                dispatch( actionFinishLoading() );
            });
    }
}

export const authEmailPasswordSignUp = ({ email, password, username }) => {
    return async (dispatch) => {
        dispatch( actionStartLoading() );

        const auth = getAuth(app);

        await createUserWithEmailAndPassword(auth, email, password)
            .then( async userCredentials => {

                const { user } = userCredentials;
                await updateProfile(user, { displayName: username });

                // dispatch( authLogin(user.uid, user.displayName) );

            }).catch( reason => {
                Swal.fire({
                    title: 'Oups! <small>Ahora mismo no podemos registrarte.</small>',
                    html: '<p>Por favor... <br/> <small>Regresa más tarde y vuelve a intentarlo ;) .</small><p>',
                    icon: 'info',
                    iconColor: 'orange',
                    confirmButtonText: 'Entiendo',
                    confirmButtonColor: 'limegreen',
                    showCloseButton: true,
                    footer: `<small>Reason Code: ${reason.code}</small>`,
                });
            }).finally(() => {
               dispatch( actionFinishLoading() );
            });
    }
}

export const authFirebaseLogout = () => {
    return async dispatch => {
        const auth = getAuth(app);

        await signOut(auth);

        dispatch( authLogout() );

        dispatch( cleanupNotesLogout() );
    }
}

export const authLogout = () => ({ type: AUTH_LOGOUT });

export const setAuthError = ( msg ) => ({
    type: AUTH_SET_ERROR,
    payload: msg
});

export const removeAuthError = () => ({ type: AUTH_REMOVE_ERROR });