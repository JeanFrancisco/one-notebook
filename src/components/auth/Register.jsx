import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import Errors from './Errors';
import { useForm } from '../../hooks/useForm';
import { authEmailPasswordSignUp, removeAuthError, setAuthError } from './actions';
import { useEffect } from 'react';

const Register = () => {

    const dispatch = useDispatch();
    const { errors } = useSelector( state => state.auth );

    useEffect(() => {
        dispatch( removeAuthError() );
    }, [ dispatch ]);

    const isFormValid = () => {

        if( validator.isEmpty(formValues.username) ) {
            dispatch( setAuthError('Su nombre de usuario no puede ser un valor vacío') );
            return false;
        }

        if( ! validator.isEmail(formValues.email) ) {
            dispatch( setAuthError('El email no es válido') );
            return false;
        }

        if( ! validator.isLength(formValues.password, { min: 6 } ) || formValues.password !== formValues.confirmpassword ) {
            dispatch( setAuthError(`Revise su contraseña. Recuerde que no puede tener menos de 6 carácteres y
            debe coincidir con la confirmación de password`) );
            return false;
        }

        if( errors !== undefined )
            dispatch( removeAuthError() );

        return true;
    }

    const onSubmit = () => {
        if( isFormValid() )
            dispatch( authEmailPasswordSignUp(formValues) );
    };

    const [ formValues, handleInputChange, handleSubmit ] = useForm({
        username: 'Francisco',
        email: 'username@example.org',
        password: '122345',
        confirmpassword: '398174',
    }, onSubmit );

    return <>
        <h4 className="auth__title">Registrar</h4>

        <Errors errors={ errors } />

        <form onSubmit={ handleSubmit }>
            <input
                type="text"
                name="username"
                className="auth__input"
                autoComplete="off"
                placeholder="Enter your username"
                onChange={ handleInputChange }
                value={ formValues.username }
                />

            <input
                type="email"
                name="email"
                className="auth__input"
                autoComplete="off"
                placeholder="Enter your e-mail"
                onChange={ handleInputChange }
                value={ formValues.email }
                />

            <input
                type="password"
                name="password"
                className="auth__input"
                placeholder="Type your password"
                onChange={ handleInputChange }
                value={ formValues.password }
                />

            <input
                type="password"
                name="confirmpassword"
                className="auth__input"
                placeholder="Please confirm your password"
                onChange={ handleInputChange }
                value={ formValues.confirmpassword }
                />

            <button
                className="btn btn-primary btn-block auth__button"
                type="submit">
                    Registrarme
            </button>
        </form>

        <Link
            className="link"
            to="/auth/login">
            ¿Ya estás registrado? Inicia Sesión.
        </Link>
    </>
}

export default Register;