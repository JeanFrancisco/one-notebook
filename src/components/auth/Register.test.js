import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { setAuthError } from './actions';
import Register from './Register';

const mockStore = configureStore([ thunk ]);

jest.mock('./actions', () => ({
    setAuthError: jest.fn().mockReturnValue({}),
    removeAuthError: jest.fn().mockReturnValue({ type: '' }),
}));

describe('Pruebas en el componente <Register/>.jsx', () => {
    const store = mockStore({
        auth: {},
    });

    store.dispatch = jest.fn();

    let wrapper;

    beforeEach(() => {
        wrapper = mount(<Provider store={ store }>
            <MemoryRouter>
                <Register/>
            </MemoryRouter>
        </Provider>);
    });

    test('Debe hacer match con el snapshot por defecto', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe mostrar el mensaje de error, si hay uno definido en el state', () => {
        const store = mockStore({
            auth: {
                errors: 'That is an error message for tests',
            }
        })

        store.dispatch = jest.fn();

        const wrapper = mount(<Provider store={ store }>
            <MemoryRouter>
                <Register />
            </MemoryRouter>
        </Provider>)

        expect( wrapper.find('.auth__errors').html() ).toContain( store.getState().auth.errors );
    });

    describe('Debe mandar el mensaje de error predefinido a la función|acción setAuthError()', () => {

        test('Caso 1: Email vacío o no valido', () => {
            wrapper.find('[type="email"]').simulate('change', {
                target: {
                    name: 'email',
                    value: ''
                }
            });

            wrapper.find('form').simulate('submit', { preventDefault() {} });

            expect( setAuthError ).toHaveBeenCalledWith('El email no es válido');
        });

        test('Caso 2: Nombre de usuario vacio', () => {
            wrapper.find('[name="username"]').simulate('change', {
                target: {
                    name: 'username',
                    value: '',
                }
            });

            wrapper.find('form').prop('onSubmit')({ preventDefault() {} });

            expect( setAuthError ).toHaveBeenCalledWith('Su nombre de usuario no puede ser un valor vacío');
        });

        test('Caso 3: Son muy cortas la Password y su Confirmacion de Password', () => {
            wrapper.find('[name="confirmpassword"]').simulate('change', {
                target: {
                    name: 'confirmpassword',
                    value: '012'
                }
            });

            wrapper.find('[name="password"]').simulate('change', {
                target: {
                    name: 'password',
                    value: '012'
                }
            });

            wrapper.find('form').prop('onSubmit')({ preventDefault() {} });

            expect( setAuthError ).toHaveBeenCalledWith(`Revise su contraseña. Recuerde que no puede tener menos de 6 carácteres y
            debe coincidir con la confirmación de password`);
        });

        test('Caso 4: Password y su Confirmacion de Password no coinciden', () => {
            wrapper.find('[name="confirmpassword"]').simulate('change', {
                target: {
                    name: 'confirmpassword',
                    value: '00000000'
                }
            });

            wrapper.find('form').prop('onSubmit')({ preventDefault() {} });

            expect( setAuthError ).toHaveBeenCalledWith(`Revise su contraseña. Recuerde que no puede tener menos de 6 carácteres y
            debe coincidir con la confirmación de password`);
        });
    })
});