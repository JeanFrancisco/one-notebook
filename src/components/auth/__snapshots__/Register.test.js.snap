// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <Register/>.jsx Debe hacer match con el snapshot por defecto 1`] = `
Array [
  <h4
    className="auth__title"
  >
    Registrar
  </h4>,
  "",
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="username"
      onChange={[Function]}
      placeholder="Enter your username"
      type="text"
      value="Francisco"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Enter your e-mail"
      type="email"
      value="username@example.org"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Type your password"
      type="password"
      value="122345"
    />
    <input
      className="auth__input"
      name="confirmpassword"
      onChange={[Function]}
      placeholder="Please confirm your password"
      type="password"
      value="398174"
    />
    <button
      className="btn btn-primary btn-block auth__button"
      type="submit"
    >
      Registrarme
    </button>
  </form>,
  <a
    className="link"
    href="/auth/login"
    onClick={[Function]}
  >
    ¿Ya estás registrado? Inicia Sesión.
  </a>,
]
`;
